File: python/one/llm.py
"""
LLM — Model adapters

Simple complete() functions for common providers.
Same interface: (prompt, system?) -> str
"""

import os
import requests


def anthropic(api_key: str | None = None, model: str = "claude-sonnet-4-20250514"):
    """Anthropic Claude adapter."""
    key = api_key or os.environ.get("ANTHROPIC_API_KEY", "")

    def complete(prompt: str, system: str | None = None) -> str:
        r = requests.post(
            "https://api.anthropic.com/v1/messages",
            headers={
                "x-api-key": key,
                "content-type": "application/json",
                "anthropic-version": "2023-06-01",
            },
            json={
                "model": model,
                "max_tokens": 4096,
                "system": system or "",
                "messages": [{"role": "user", "content": prompt}],
            },
            timeout=60,
        )
        r.raise_for_status()
        return r.json()["content"][0]["text"]

    return complete


def openai(api_key: str | None = None, model: str = "gpt-4o"):
    """OpenAI adapter."""
    key = api_key or os.environ.get("OPENAI_API_KEY", "")

    def complete(prompt: str, system: str | None = None) -> str:
        r = requests.post(
            "https://api.openai.com/v1/chat/completions",
            headers={
                "Authorization": f"Bearer {key}",
                "content-type": "application/json",
            },
            json={
                "model": model,
                "messages": [
                    {"role": "system", "content": system or ""},
                    {"role": "user", "content": prompt},
                ],
            },
            timeout=60,
        )
        r.raise_for_status()
        return r.json()["choices"][0]["message"]["content"]

    return complete

File: templates/web/src/lib/llm.ts
import type { Env, Message } from './types'

const DEFAULT_MODEL = 'meta-llama/llama-4-maverick'

export async function chat(
  env: Env,
  systemPrompt: string,
  messages: Message[],
  model?: string
): Promise<string> {
  const modelId = model ?? env.AGENT_MODEL ?? DEFAULT_MODEL

  const isGroq = modelId.startsWith('groq/') && env.GROQ_API_KEY
  const url = isGroq
    ? 'https://api.groq.com/openai/v1/chat/completions'
    : 'https://openrouter.ai/api/v1/chat/completions'

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${isGroq ? env.GROQ_API_KEY : env.OPENROUTER_API_KEY}`,
  }

  if (!isGroq) {
    headers['HTTP-Referer'] = 'https://one.ie'
    headers['X-Title'] = 'ONE-Demo'
  }

  const res = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      model: isGroq ? modelId.slice(5) : modelId,
      messages: [{ role: 'system', content: systemPrompt }, ...messages],
      max_tokens: 1024,
      temperature: 0.7,
    }),
  })

  if (!res.ok) throw new Error(`LLM error: ${await res.text()}`)

  const data = (await res.json()) as { choices?: { message?: { content?: string } }[] }
  return data.choices?.[0]?.message?.content ?? ''
}
